src/did/src/sign_strategy.rs

```diff
@@ -0,0 +1,334 @@
+use std::borrow::Cow;
+use std::cell::RefCell;
+
+use async_trait::async_trait;
+use candid::CandidType;
+use eth_signer::ic_sign::{IcSigner, SigningKeyId};
+use eth_signer::{Signer, Wallet};
+use ethers_core::k256::ecdsa::SigningKey;
+use ethers_core::types::transaction::eip2718::TypedTransaction;
+use ethers_core::utils;
+use ic_exports::ic_ic00_types::DerivationPath;
+use ic_stable_structures::Storable;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::error::{EvmError, Result};
+use crate::transaction::Signature;
+use crate::H160;
+
+/// A trait that abstracts out the transaction signing component
+#[async_trait(?Send)]
+pub trait TransactionSigner {
+    /// Returns the `sender` address for the given identity
+    async fn get_address(&self) -> Result<H160>;
+
+    /// Signe the created transaction
+    async fn sign_transaction(&self, transaction: &TypedTransaction) -> Result<Signature>;
+}
+
+/// Signing strategy for signing EVM transactions
+#[derive(CandidType, Clone, Debug, Deserialize)]
+pub enum SigningStrategy {
+    /// Local signing key
+    Local { private_key: [u8; 32] },
+    /// Use management canister and ECDSA signing endpoints
+    ManagementCanister {
+        key_id: SigningKeyId,
+        derivation_path: DerivationPath,
+    },
+}
+
+impl SigningStrategy {
+    /// Create signing object from the current strategy
+    pub fn make_signer(self, chain_id: u64) -> Result<TxSigner> {
+        match self {
+            SigningStrategy::Local { private_key } => {
+                let signer = SigningKey::from_slice(&private_key).map_err(|e| {
+                    EvmError::from(format!("failed to deserialize signing key: {e}"))
+                })?;
+                let address = utils::secret_key_to_address(&signer);
+                let wallet = Wallet::new_with_signer(Cow::Owned(signer), address, chain_id);
+                Ok(TxSigner::Local(LocalTxSigner::new(wallet)))
+            }
+            SigningStrategy::ManagementCanister {
+                key_id,
+                derivation_path,
+            } => Ok(TxSigner::ManagementCanister(ManagementCanisterSigner::new(
+                key_id,
+                derivation_path,
+            ))),
+        }
+    }
+}
+
+/// Transaction signer
+#[derive(Serialize, Deserialize, Clone)]
+pub enum TxSigner {
+    Local(LocalTxSigner),
+    ManagementCanister(ManagementCanisterSigner),
+}
+
+impl Default for TxSigner {
+    fn default() -> Self {
```

Comment:
I would remove this default implementation as it is quite dangerous the way it is implemented. The user could not notice that it is signing with an extremely predictable key.

Reply:
The default implementation was removed.
In the minter canister we use now ICSigner for a test network as the default value which is safe